hack/upstream-features/upload.go

```diff
@@ -151,41 +173,67 @@ func ensureFeatureProduct(syncedFeature syncedFeature) error {
 			},
 		},
 		Metadata: map[string]string{
-			licenseapi.MetadataKeyProductForFeature: syncedFeature.name,
+			licenseapi.MetadataKeyProductForFeature: syncedFeature.Feature.LookupKey,
 		},
 	})
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	_, err = productfeature.New(&stripe.ProductFeatureParams{
+	_, err = stripeClient.ProductFeatures.New(&stripe.ProductFeatureParams{
 		Product:            &product.ID,
-		EntitlementFeature: &syncedFeature.stripeID,
+		EntitlementFeature: &syncedFeature.Feature.ID,
 	})
 	if err != nil {
-		return err
+		return nil, err
 	}
-	return nil
+	return product, nil
 }
 
-func ensureAttachAll(featureIDs map[string]syncedFeature) error {
-	productSearch := stripeproducts.Search(&stripe.ProductSearchParams{
+func EnsureAttachAll(
+	stripeClient *stripeclient.API,
+	features map[string]SyncedFeature,
+	attachTo []stripe.Product,
+) error {
+	productSearch := stripeClient.Products.Search(&stripe.ProductSearchParams{
 		SearchParams: stripe.SearchParams{
-			Query: fmt.Sprintf(metadataQueryFmt, licenseapi.MetadataKeyAttachAll, "true"),
+			Query: fmt.Sprintf(metadataQueryFmt, licenseapi.MetadataKeyAttachAll, licenseapi.MetadataValueTrue),
 		},
 	})
 	if err := productSearch.Err(); err != nil {
 		return err
 	}
 	for productSearch.Next() {
 		prod := productSearch.Product()
-		featuresToCheck := maps.Clone[map[string]syncedFeature](featureIDs)
+		attachTo = append(attachTo, *prod)
+	}
 
-		if err := SearchProductForFeatures(prod.ID, featuresToCheck); err != nil {
+	prodIDs := map[string]bool{}
+
+	for _, prod := range attachTo {
+		alreadyDone, ok := prodIDs[prod.ID]
+		if ok && alreadyDone {
+			continue
+		}
+		prodIDs[prod.ID] = true
+
+		featuresToCheck := maps.Clone(features)
+
+		if err := SearchProductForFeatures(stripeClient, prod.ID, featuresToCheck); err != nil {
 			return err
 		}
-		for featureID := range featuresToCheck {
-			_, err := productfeature.New(&stripe.ProductFeatureParams{
+		for featureID, feat := range featuresToCheck {
+			metadataHiddenValue, ok := feat.Feature.Metadata[licenseapi.MetadataKeyFeatureIsHidden]
+			if ok || metadataHiddenValue == licenseapi.MetadataValueTrue {
```

Comment:
This is redundant, if if `metadataHiddenValue == licenseapi.MetadataValueTrue`, then ok is true. If ok is true then the evaluation short circuits and skips the second condition.

You probably meant to use "&&" here, but even then ok is unnecessary as metadataHiddenValue will be "".

```suggestion
			if feat.Feature.Metadata[licenseapi.MetadataKeyFeatureIsHidden] == licenseapi.MetadataValueTrue
```